File: AndroidClipSync/app/src/main/java/com/mtechdwork/clipsync/TCPHandler.java
package com.mtechdwork.clipsync;

import android.content.Context;
import android.util.Log;

import org.json.JSONObject;

import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.net.InetAddress;
import java.net.ServerSocket;
import java.net.Socket;
import java.nio.charset.StandardCharsets;

public class TCPHandler extends Thread {

    private boolean running = true;
    private final Context context;
    private ServerSocket serverSocket;

    TCPHandler(Context context) {
        this.context = context;
    }

    public void stopRunning() {
        try {
            running = false;
            if (serverSocket != null) serverSocket.close();
        } catch (Exception e) {
            log(e.getMessage(), 2);
        }
    }

    private void log(String message, int type) { // Debug method
        // Type: 0 - Info, 1 - Warning, 2 - Error
        boolean debug = false;
        if (!debug) return;
        String className = "[TCP Handler]";
        switch (type) {
            case 0:
                Log.i(className, message);
                break;

            case 1:
                Log.w(className, message);
                break;

            case 2:
                Log.e(className, message);
        }
    }

    private void clientHandler(InetAddress ipAddress, String encryptData) {
        try {
            byte[] associatedData = "metadata".getBytes(StandardCharsets.UTF_8);
            String jsonMessage = XChaChaCrypto.decrypt(encryptData, associatedData);

            JSONObject jsonObject = new JSONObject(jsonMessage);

            switch (jsonObject.getString("type")) {
                case "otp_auth":
                    Authenticator authenticator = new Authenticator(context);
                    String otp = jsonObject.getString("value");
                    if (authenticator.checkOTPMatch(otp)) {
                        Communication communication = new Communication(context);
                        communication.sendClipboardData(ipAddress);
                    }
                    break;

                case "data":
                    ClipboardData.writeClipboard(context, jsonObject.getString("value"));
                    break;
            }
        } catch (Exception e) {
            log(e.getMessage(), 2);
        }
    }

    @Override
    public void run() {
        try {
            int PORT = 7071;
            serverSocket = new ServerSocket(PORT);
            while (running) {
                Socket clientSocket = serverSocket.accept(); // Chờ client kết nối
                log("Client connected: " + clientSocket.getInetAddress(), 0);

                // Đọc dữ liệu từ client
                BufferedReader reader = new BufferedReader(new InputStreamReader(clientSocket.getInputStream()));
                String message = reader.readLine();
                log("Received: " + message, 0);

                // Handler
                InetAddress clientIP = clientSocket.getInetAddress();
                clientHandler(clientIP, message);

                // Close connection
                clientSocket.close();
            }
        } catch (Exception e) {
            log(e.getMessage(), 2);
        }
    }

    @Override
    public void interrupt() {
        super.interrupt();
        log("Thread interrupt!", 0);
        stopRunning();
    }
}

File: WindowsClipSync/ClipSync/Classes/TCPHandler.vb
Imports System.IO
Imports System.Net
Imports System.Net.Sockets
Imports System.Text
Imports System.Threading
Imports Newtonsoft.Json.Linq

Public Class TCPHandler
	Private tcpListening As Thread
	Private tcpListener As TcpListener
	Private running = False
	Private TCP_PORT = 7071

	Public Sub startListening()
		running = True
		tcpListening = New Thread(AddressOf listening)
		tcpListening.IsBackground = True
		tcpListening.Start()
	End Sub

	Public Sub stopListening()
		running = False
		If tcpListener IsNot Nothing Then
			tcpListener.Stop()
		End If
	End Sub

	Private Sub clientHandler(ipAddr As IPAddress, encryptData As String)
		Try
			Dim jsonMessage = XChaChaCrypto.decrypt(encryptData, Encoding.UTF8.GetBytes("metadata"))

			ClipSyncDebug.log("[TCPHandler] Message after decrypt: " + jsonMessage)

			Dim jsonObject = JObject.Parse(jsonMessage)

			Select Case jsonObject.GetValue("type").ToString()
				Case "otp_auth"
					Dim otp As String = jsonObject.GetValue("value").ToString()

					If Authenticator.checkOTPMatch(otp) Then
						Communication.sendClipboardData(ipAddr)
					End If

				Case "data"
					ClipboardData.writeClipboard(jsonObject.GetValue("value").ToString)
			End Select
		Catch ex As Exception
			ClipSyncDebug.log("[TCPHandler] Exception: " + ex.Message)
		End Try
	End Sub

	Private Sub listening()
		Try
			Dim ipAddr As IPAddress = LocalIP.getIPAddress()
			tcpListener = New TcpListener(ipAddr, TCP_PORT)
			tcpListener.Start()

			While running
				Using client As Socket = tcpListener.AcceptSocket
					Dim clientIP As IPAddress = IPAddress.Parse(CType(client.RemoteEndPoint, IPEndPoint).Address.ToString())
					ClipSyncDebug.log("[TCPHandler] Client connected: " + clientIP.ToString)

					' Read data from client
					Using stream As NetworkStream = New NetworkStream(client)
						Using reader As New StreamReader(stream)
							Dim message As String = reader.ReadLine()
							ClipSyncDebug.log("[TCPHandler] Received: " & message)

							' Handler
							clientHandler(clientIP, message)
						End Using
					End Using

					' Close connection
					client.Close()
				End Using
			End While

			If tcpListener IsNot Nothing Then
				tcpListener.Stop()
			End If
		Catch ex As Exception
			ClipSyncDebug.log("[TCPHandler] Exception: " + ex.Message)
		End Try
	End Sub
End Class
